_posts/2023-08-16-debugging-android-devices.md

```diff
@@ -0,0 +1,277 @@
+[ 1462.184957] lr : _raw_spin_lock+0x48/0x50
+[ 1462.184958] sp : ffff00000801bd70
+[ 1462.184961] x29: ffff00000801bd80 x28: ffff80007df0d068 
+[ 1462.184966] x27: ffff80007df0d040 x26: 000001546d61b824 
+[ 1462.184970] x25: ffff80007df0d050 x24: ffff0000081412ec 
+[ 1462.184975] x23: ffff000008cf0e80 x22: ffff000008e1db50 
+[ 1462.184979] x21: 0000000000000003 x20: 00610062003a0072 
+[ 1462.184983] x19: ffff80006cd0fe80 x18: 0000000081005000 
+[ 1462.184988] x17: 000000007571732f x16: 0000000000000000 
+[ 1462.184992] x15: ffff000008e1db50 x14: ffff0000089e3378 
+[ 1462.184997] x13: ffff000008cf1ad0 x12: 0000000000100000 
+[ 1462.185001] x11: ffff80007df10ac8 x10: ffff80007df10ac0 
+[ 1462.185005] x9 : ffff000008cf1ac0 x8 : ffff80006cd0fe80 
+[ 1462.185010] x7 : ffff000013203c70 x6 : ffff000013203be4 
+[ 1462.185014] x5 : ffff000013203c64 x4 : 0000000000000004 
+[ 1462.185018] x3 : 0000000000000001 x2 : 0000000000000001 
+[ 1462.185022] x1 : 000000007571732f x0 : ffff80006cd0fe80 
+[ 1462.185027] Call trace:
+[ 1462.185033]  queued_spin_lock_slowpath+0x104/0x284
+[ 1462.185039]  scheduler_tick+0x48/0x100
+[ 1462.185045]  update_process_times+0x80/0x98
+[ 1462.185052]  tick_sched_timer+0x90/0xf0
+[ 1462.185056]  __hrtimer_run_queues+0x118/0x1bc
+[ 1462.185060]  hrtimer_interrupt+0xf0/0x3e0
+[ 1462.185067]  arch_timer_handler_virt+0x34/0x40
+[ 1462.185073]  handle_percpu_devid_irq+0x7c/0x154
+[ 1462.185077]  __handle_domain_irq+0x7c/0xbc
+[ 1462.185082]  gic_handle_irq+0x4c/0xb8
+[ 1462.185086]  el1_irq+0xe8/0x190
+[ 1462.185091]  arch_cpu_idle+0x10/0x18
+[ 1462.185097]  do_idle+0xbc/0x268
+[ 1462.185101]  cpu_startup_entry+0x20/0x24
+[ 1462.185108]  secondary_start_kernel+0x14c/0x158
+[ 1462.185115] Code: 8b2e4dee 120005ad f85f81ce 8b2d512d (f82d69ca) 
+[ 1462.185119] ---[ end trace 9aea57dd6aa91794 ]---
+[ 1462.235526] Kernel panic - not syncing: Fatal exception in interrupt
+[ 1462.235539] SMP: stopping secondary CPUs
+[ 1462.235550] Kernel Offset: disabled
+[ 1462.235557] CPU features: 0x0,24002004
+[ 1462.235560] Memory Limit: none
+[ 1462.492479] Rebooting in 5 seconds..
+```
+
+Finally, we have the actual crash itself. Detailed information about the process that crashed is provided such as the name of the binary, what CPU it was running on, the call trace, and register values. This can all be incredibly helpful in understand what is happening in the kernel at the time of the panic. You'll notice here as well that call trace has memory offsets for each of the function calls. With the use of `addr2line` ([man page](https://man7.org/linux/man-pages/man1/addr2line.1.html)), you are able to get the filename and line numbers associated with those offsets.
+
+### Application Not Responding (ANRs)
+
+Any time the UI thread of an Android Application is excessively blocked, Android will trigger an ANR. This presents a dialogue in the foreground allowing the user to forcibly close the application.
```

Comment:
Late changes but:

Any time the UI thread of an Android _device_ 

This presents a _dialog_